src/libraries/System.IO.Compression/tests/ZipArchive/zip_UpdateTests.cs

```diff
@@ -562,6 +751,106 @@ public void Update_PerformMinimalWritesWithDataAndHeaderChanges(int dataChangeIn
             }
         }
 
+        //[Theory]
+        //[InlineData(0, 0)]
+        //[InlineData(20, 40)]
+        //[InlineData(30, 10)]
+        //public async Task Update_PerformMinimalWritesWithDataAndHeaderChanges_Async(int dataChangeIndex, int lastWriteTimeChangeIndex)
+        //{
+        //    byte[] sampleEntryContents = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
+        //    byte[] sampleZipFile = await CreateZipFile(50, sampleEntryContents, async: true);
+        //    byte[] expectedUpdatedEntryContents = [19, 18, 17, 16, 15, 14, 13, 12, 11, 10];
+
+        //    await using (MemoryStream ms = new MemoryStream())
+        //    {
+        //        await ms.WriteAsync(sampleZipFile);
+        //        ms.Seek(0, SeekOrigin.Begin);
+
+        //        await using (CallTrackingStream trackingStream = new CallTrackingStream(ms))
+        //        {
+        //            // Open the archive in Update mode, then rewrite the data of the {dataChangeIndex}th entry
+        //            // and set the LastWriteTime of the {lastWriteTimeChangeIndex}th entry.
+        //            // Verify the correct number of writes performed as a result, then reopen the same
+        //            // archive, get the entries and make sure that the fields hold the expected value.
+        //            int writesCalled = trackingStream.TimesCalled(nameof(trackingStream.Write));
+        //            int writeBytesCalled = trackingStream.TimesCalled(nameof(trackingStream.WriteByte));
+        //            ZipArchive target = await ZipArchive.CreateAsync(trackingStream, ZipArchiveMode.Update, leaveOpen: true, entryNameEncoding: null);
```

Comment:
@edwardneal This new test, and the other 2 new ones above are currently commented because they fail. I'm investigating. They are heavily involving the recent changes.

Reply:
I see what you mean. In principle, it's because we're measuring the number of calls to the synchronous Write/WriteByte methods, but the async path won't use them. There's apparently one synchronous call to Write/WriteByte though, which is a second problem. I'll take a look at this.

Reply:
There are three classes inheriting from Stream in ZipCustomStreams.cs: `WrappedStream`, `SubReadStream` and `CheckSumAndSizeWriteStream`. We need to implement `DisposeAsync` on each of these - the default implementation indirectly calls the synchronous `Write` method.

Once that's done, all of the tests can be uncommented. The `_Async` versions just need to refer to `trackingStream.TimesCalled(nameof(trackingStream.WriteAsync))` rather than `nameof(trackingStream.Write)`.

The tests also track the number of times WriteByte is called, and this doesn't have an async counterpart. I don't think we need to track this, it's never called.

Reply:
Fixed the problem. I was having two problems:
- The main issue was that I needed your two recent fixes, there was some weird behavior without them.
- I learned I can't just be playing with pre-existing tests that create FileStreams without passing `useAsync` to their constructor and then expect them to behave nicely in all cases. I made sure all the tests requiring FileStreams were using async.